### ISSUE TITLE
Bug in deleteAll in Maps

### ISSUE THREAD

**Issue by hiranpodi** [immutable-js/immutable-js#1701]
This issue is with Maps have 15 or more elements, 

How to reproduce:

1. Create a immutable Map with 15 elements.  e.g.

>  ( {a: 'aaaa', b: 'bbbb'...... p: 'pppp'})

2. Add two of the Keys to an array.
     e.g

> . [ 'a', 'b']

3. Use "deleteAll" to delete elements with keys selected in step 2. 
     e.g 

> map1.deleteAll([ 'a', 'b'])

4. Add two new elements to the Map. 
     e.g.

>  map1.set('x','xxx'); 
>             map1.set('y','yyy');

5. Repeat step "2" to "4" , 15 times. on the 16th deletion the two elements will not be deleted from the Map.

A working example of this issue can be found in this link:

[http://jsfiddle.net/n8e9ufgc/4](http://jsfiddle.net/n8e9ufgc/4)

Version : 4.0.0.-rc.12

![image](https://user-images.githubusercontent.com/13897515/55209789-6ead2d00-5238-11e9-9bf1-981824d1d774.png)

**Comment by acusti** [comment]
i reproduced this and have a few additional findings:

1. it only applies to `OrderedMap`, not `Map`: [fiddle with `OrderedMap`](http://jsfiddle.net/69r8whgf/6/), [fiddle with `Map`](http://jsfiddle.net/69r8whgf/7/)
2. only a single new item needs to be set and a single old item removed via `deleteAll` to trigger the bug
3. the point at which the `deleteAll` calls start failing varies depending on the size of the `OrderedMap` and on how many items are being added and deleted. having an OrderedMap of size 15 will fail quickest (after 18 iterations), while a size of 14 and a size of 16 will take 19 iterations. try this in the browser console at https://immutable-js.github.io/immutable-js/docs/ to see that happen, then try tweaking the top config options:
```js
var collectionSize = 15;
var totalIterations = 18;

var map = Immutable.OrderedMap();
var clickCount = 0;
for (var i = 0; i < collectionSize - 1; i++) { 
  map = map.set(i, {a: i})
}

var previousIndex = collectionSize - 1;
var index;

for (var i = 0; i < totalIterations; i++) {
  // First, add a new item
  index = Math.random() * 100;
  map = map.set(index, {a: index});
  // Then delete the previously added item
  map = map.deleteAll([previousIndex]);
  // Update previousIndex for next iteration
  previousIndex = index;
}
map;
```
by the end of the above code snippet, `map.size` will be `16` when it should be `15`. if you run the same thing but have the second loop execute one less iteration (totalIterations = 17), the bug will not exhibit and `map.size` will be `15`.

the amount of time it takes to trigger the bug seems to remain symmetrical from the distance away from size 15. when collectionSize = 1, it takes 32 iterations, which is the same for collectionSize = 29.

**Comment by jdeniau** [comment]
Hi,

This has been fixed in the latest rc `4.0.0-rc.14`

Thank you for your patience !
